EventFilter/L1TRawToDigi/plugins/implementations_stage2/BMTFPackerInputs.cc

```diff
@@ -0,0 +1,367 @@
+#include "BMTFPackerInputs.h"
+
+#include <vector>
+//#include <bitset>//debug
+
+
+namespace l1t 
+{
+   namespace stage2 
+   {
+
+      const int BMTFPackerInputs::ownLinks_[]={4,5,12,13,20,21,22,23,28,29};
+
+      Blocks BMTFPackerInputs::pack(const edm::Event& event, const PackerTokens* toks)
+      {
+	int board_id = (int)board();
+
+        auto muonPhToken = static_cast<const BMTFTokens*>(toks)->getInputMuonTokenPh();
+        auto muonThToken = static_cast<const BMTFTokens*>(toks)->getInputMuonTokenTh();
+
+        Blocks blocks;
+  
+        edm::Handle<L1MuDTChambPhContainer> phInputs;
+        event.getByToken(muonPhToken, phInputs);
+        edm::Handle<L1MuDTChambThContainer> thInputs;
+        event.getByToken(muonThToken, thInputs);
+
+	uint32_t qualEta_32bit = 0;
+	uint32_t posEta_0_32bit = 0;	
+	uint32_t posEta_n2_32bit = 0, posEta_n1_32bit = 0;
+	uint32_t posEta_p2_32bit = 0, posEta_p1_32bit = 0;
+
+	bool moreBXeta = false;
+	for (int link = 0; link <= 35; link++) {
+	  //	  std::cout << "link : " << link << std::endl;
+
+	  if ( (link >= 6 && link < 8) ||
+	       (link >= 14 && link < 16) ||
+	       (link >= 30 && link < 32) ) 
+	    continue;
+
+	  //initializing null block_payloads and block_id
+	  std::vector<uint32_t> payload_0 (6,0);
+	  std::vector<uint32_t> payload_p1(6,0);
+	  std::vector<uint32_t> payload_n1(6,0);
+	  std::vector<uint32_t> payload_p2(6,0);
+	  std::vector<uint32_t> payload_n2(6,0);
+
+	  unsigned int block_id = 2*link;
+	  
+	  std::vector<bool> bxPresent(5,false);
+	  bool moreBXphi = false;
+
+	  unsigned int BC = 0;
+	  
+	  //The first 4 phi words for the link's payload
+
+          for (const auto &  iphi : *(phInputs->getContainer()) )
+	  {   
+	      if (iphi.bxNum() != 0)
+		moreBXphi = true;
+
+	      //	      std::cout << "scNum+1 = " << iphi->scNum()+1 << ",   board_id = " << board_id << std::endl;
+	      // 	      std::cout << "bx, station = " << iphi->bxNum() << ", " << iphi->stNum() << std::endl;
+
+	      //BC = iphi->BxCnt();//this thing here is not completely functional
+
+	      if ( iphi.scNum()+1 != board_id )
+		continue;
+	      //	      std::cout << "correct board" << std::endl;
+
+	      if (link != ownLinks_[4+2*(iphi.whNum())+iphi.Ts2Tag()])
+		continue;
+	      // 	      std::cout << "correct link" << std::endl;
+	     
+	      bxPresent[2+iphi.bxNum()] = true;
```

Comment:
At this line, some Test workflows are crashing
eg https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20834/23919/runTheMatrix-results/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017/step2_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+ALCAFull_2017+HARVESTFull_2017.log


`iphi.bxNum() `gets value  -3, so
`2+iphi.bxNum() = -1`

... not legal index value.  A protection is needed here for sure.

But also need feedback from experts if this bxNum() = -3 is expected, 
and if this case one should simply skip the Ph DT Segment?
Attn @panoskatsoulis @gkaratha

Reply:
@rekovic @gkaratha 
You are right Vladimir, this value is not expected by the module.
The reason is that the packer is not designed to pack information that comes from bx greater than 2 or lower than -2.

I've managed to run it with a protection, yes.
But the real question is that, what do we need?
-> the packer to be able pack only what the bmtf subsystem is able to give us
(this means that the protection is enough and the RAW will be realistic)
OR
-> the packer to be able to pack everything form the generated MC.
(this means that the module needs extension but the RAW won't be realistic.)

Reply:
Implemented the protection.
We are only packing DT inputs for bx in [-2,2], as the case if fw.